openmdao/core/tests/test_group.py

```diff
@@ -96,6 +96,123 @@ def configure(self):
 
 class TestGroup(unittest.TestCase):
 
+    def test_promotes_outputs_in_config(self):
+
+        class SimpleGroup(om.Group):
+
+            def setup(self):
+
+                self.add_subsystem('comp1', om.IndepVarComp('x', 5.0))
+                self.add_subsystem('comp2', om.ExecComp('b=2*a'))
+
+            def configure(self):
+                self.promotes('comp2', outputs=['b'])
+
+        top = om.Problem(model=SimpleGroup())
+        top.setup()
+
+        self.assertEqual(top['b'], 1)
+        with self.assertRaises(UnboundLocalError) as cm:
+            top['a']
+
+        self.assertEqual(str(cm.exception),
+                         "local variable 'val' referenced before assignment")
+
+    def test_promotes_inputs_in_config(self):
+
+        class SimpleGroup(om.Group):
+
+            def setup(self):
+
+                self.add_subsystem('comp1', om.IndepVarComp('x', 5.0))
+                self.add_subsystem('comp2', om.ExecComp('b=2*a'))
+
+            def configure(self):
+                self.promotes('comp2', inputs=['a'])
+
+        top = om.Problem(model=SimpleGroup())
+        top.setup()
+
+        self.assertEqual(top['a'], 1)
+        with self.assertRaises(UnboundLocalError) as cm:
+            top['b']
+
+        self.assertEqual(str(cm.exception),
+                         "local variable 'val' referenced before assignment")
+
+    def test_promotes_in_config(self):
+
+        class SimpleGroup(om.Group):
+
+            def setup(self):
+
+                self.add_subsystem('comp1', om.IndepVarComp('x', 5.0))
+                self.add_subsystem('comp2', om.ExecComp('b=2*a'))
+
+            def configure(self):
+                self.promotes('comp1', any=['*'])
+
+        top = om.Problem(model=SimpleGroup())
+        top.setup()
+
+        self.assertEqual(top['x'], 5)
+        with self.assertRaises(UnboundLocalError) as cm:
+            top['a']
+
+        self.assertEqual(str(cm.exception),
+                         "local variable 'val' referenced before assignment")
+
+    def test_multiple_promotes(self):
```

Comment:
I don't see any tests in here that do promotes with aliasing.  We should probably have a couple of those, one with multiple nested groups that does promotion with aliasing successfully and another that maybe tries and fails to promote an aliased child name using the original name.

Reply:
Added